### PR TITLE
replaceVarsWith: init

### PR DESCRIPTION
**Pull request by wolfgangwalther**
Working my way towards removing substituteAll in https://github.com/NixOS/nixpkgs/issues/237216.

This takes the extended features of nix substituteAll to a replaceVars variant to get rid of those cases that use substituteAll to build a full package with meta information etc.

There are at least 15 more packages which will make use of `replaceVarsWith`, but I didn't include the changes here to keep the number of notifications lower while we iterate on the function itself. All those other cases that I currently know of use `isExecutable = true`.

Part of my overall effort for structuredAttrs in https://github.com/NixOS/nixpkgs/issues/205690.

@philiptaron @emilazy

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by philiptaron** [comment]
Welcome to the committers, @wolfgangwalther! Glad to have you here, and thanks as always for your high quality and well-thought-out PRs and collaboration.

**Review by philiptaron — COMMENTED** [review]
I should be able to give this a thorough review this week. Thank you for your patience with the silence so far, @wolfgangwalther.

The interface I expected for `replaceVarsWith` was an [adapted `stdenv.mkDerivation`](https://github.com/NixOS/nixpkgs/pull/234651), rather than a function of three arguments. I'm open to a function of three arguments, but I need to wrap my head around it.

The short version I had in mind was:

```
{ replaceVarsWith }:

path: attrs:

replaceVarsWith {
  src = path;
  replacements = attrs;
}
```

It's sort of fundamentally odd to me for a `With` variant to have anything other than one attrset as its input. 

Regarding #360576, it's sort of a brute-force approach, and I don't like it as much as I like a general-purpose `replaceVarsWith`.

File: pkgs/build-support/replace-vars/default.nix [hunk 1]
```diff
@@ -52,35 +54,50 @@ let
...
+        if grep -Pqe ${regex} "$out"; then
+          echo The following look like unsubstituted Nix identifiers that remain in "$out":
+          grep -Poe ${regex} "$out"
+          echo Use the more precise '`substitute`' function if this check is in error.
+          exit 1
+        fi
+        runHook postCheck
+      '';
+  }
+  // drvAttrs
```
[the remark this answers is not in the document]
Comment (philiptaron) [reply pkgs/build-support/replace-vars/default.nix:102]: Yeah. I'm pretty amenable to that, combined with judicious `lib.assert` for attempts to provide a `buildPhase`, `checkPhase`, etc.

File: pkgs/build-support/replace-vars/short.nix
Comment (philiptaron) [inline pkgs/build-support/replace-vars/short.nix:1]: In order to keep [the documentation](https://noogle.dev/f/pkgs/replaceVars) we'll have to copy a lot of text. Let's do that. Documentation is usually repetitive for wrappers, and this function deserves to have good documentation, IMO.

File: pkgs/build-support/replace-vars/default.nix [hunk 2]
```diff
@@ -45,42 +59,66 @@ let
...
-      if grep -Pqe ${regex} "$out"; then
-        echo The following look like unsubstituted Nix identifiers that remain in "$out":
-        grep -Poe ${regex} "$out"
-        echo Use the more precise '`substitute`' function if this check is in error.
-        exit 1
+  optionalAttrs =
+    if (builtins.intersectAttrs attrs forcedAttrs == { }) then
+      builtins.removeAttrs attrs [ "replacements" ]
+    else
+      throw "Passing any of ${builtins.concatStringsSep ", " (builtins.attrNames forcedAttrs)} to replaceVarsWith is not supported.";
```
Comment (philiptaron) [inline pkgs/build-support/replace-vars/default.nix:72]: Amazing. It's like a typed language, almost. 😁

**Review by philiptaron — CHANGES_REQUESTED** [review]
I like it *a lot*. One request for the copy-and-paste machine to come out, then I'd like to merge. This looks great in all the users.

**Comment by philiptaron** [comment]
If the PR is **just** the `replaceVarsWith`, is it able to target master?

**Review by philiptaron — APPROVED** [review]

**Comment by philiptaron** [comment]
Looks great, tastes great. Want to use your new committer powers to push the button?

**Comment by baloo** [comment]
I believe this broke any call to `replaceVars` as it still references the old path in `all-packages.nix`

**Pull request merged**